llama-index-core/llama_index/core/agent/multi_agent/multi_agent_workflow.py

```diff
@@ -0,0 +1,577 @@
+                )
+
+    async def _call_llm(
+        self,
+        ctx: Context,
+        llm: LLM,
+        llm_input: List[ChatMessage],
+        tools: List[AsyncBaseTool],
+        mode: AgentMode,
+    ) -> AgentOutput:
+        """Call the LLM with the given input and tools."""
+        if mode == AgentMode.DEFAULT:
+            if llm.metadata.is_function_calling_model:
+                return await self._call_function_calling_agent(
+                    ctx, llm, llm_input, tools
+                )
+            else:
+                return await self._call_react_agent(ctx, llm, llm_input, tools)
+        elif mode == AgentMode.REACT:
+            return await self._call_react_agent(ctx, llm, llm_input, tools)
+        elif mode == AgentMode.FUNCTION_CALLING:
+            return await self._call_function_calling_agent(ctx, llm, llm_input, tools)
+        else:
+            raise ValueError(f"Invalid agent mode: {mode}")
+
+    @step
+    async def init_run(self, ctx: Context, ev: StartEvent | AgentOutput) -> AgentInput:
+        """Sets up the workflow and validates inputs."""
+        if isinstance(ev, StartEvent):
+            await self._init_context(ctx)
+
+            user_msg = ev.get("user_msg")
+            chat_history = ev.get("chat_history")
+            if user_msg and chat_history:
+                raise ValueError("Cannot provide both user_msg and chat_history")
+
+            if isinstance(user_msg, str):
+                user_msg = ChatMessage(role="user", content=user_msg)
+
+            await ctx.set("user_msg_str", user_msg.content)
+
+            # Add messages to memory
+            memory: BaseMemory = await ctx.get("memory")
+            if user_msg:
+                await memory.aput(user_msg)
+                input_messages = memory.get(input=user_msg.content)
+
+                # Add the state to the user message if it exists and if requested
+                current_state = await ctx.get("current_state")
+                if self.state_prompt and current_state:
+                    user_msg.content = self.state_prompt.format(
+                        state=current_state, msg=user_msg.content
+                    )
+
+                await memory.aput(user_msg)
+            else:
+                memory.set(chat_history)
+                input_messages = memory.get()
+        else:
+            user_msg_str = await ctx.get("user_msg_str")
+            memory: BaseMemory = await ctx.get("memory")
+            input_messages = memory.get(input=user_msg_str)
+
+        # send to the current agent
+        current_agent = await ctx.get("current_agent")
+        return AgentInput(input=input_messages, current_agent=current_agent)
+
+    @step
+    async def setup_agent(self, ctx: Context, ev: AgentInput) -> AgentSetup:
+        """Main agent handling logic."""
+        agent_config: AgentConfig = (await ctx.get("agent_configs"))[ev.current_agent]
+        llm_input = ev.input
+
+        # Setup the tools
```

Comment:
*sobbing* set... up... please... not "setup"...

Reply:
🙃